deepspeed/launcher/runner.py

```diff
@@ -356,30 +358,34 @@ def main(args=None):
 
     resource_pool = fetch_hostfile(args.hostfile)
 
-    # respect CUDA_VISIBLE_DEVICES for a single node and no explicit resource filters
-    cuda_visible_devices = os.environ.get("CUDA_VISIBLE_DEVICES", "")
-    if not resource_pool and len(cuda_visible_devices):
-        detected_str = f"Detected CUDA_VISIBLE_DEVICES={cuda_visible_devices}"
-        if len(args.include) or len(
-                args.exclude) or args.num_nodes > 1 or args.num_gpus > 0:
-            print(
-                f"{detected_str} but ignoring it because one or several of --include/--exclude/--num_gpus/--num_nodes cl args were used. If you want to use CUDA_VISIBLE_DEVICES don't pass any of these arguments to deepspeed."
-            )
-        else:
-            args.include = f"localhost:{cuda_visible_devices}"
-            print(f"{detected_str}: setting --include={args.include}")
-        del os.environ["CUDA_VISIBLE_DEVICES"]
-
-    if args.num_nodes >= 0 or args.num_gpus >= 0:
-        if args.include != "" or args.exclude != "":
-            raise ValueError("Cannot specify num_nodes/gpus with include/exclude")
+    if torch.cuda.is_available():
+        # respect CUDA_VISIBLE_DEVICES for a single node and no explicit resource filters
+        cuda_visible_devices = os.environ.get("CUDA_VISIBLE_DEVICES", "")
+        if not resource_pool and len(cuda_visible_devices):
+            detected_str = f"Detected CUDA_VISIBLE_DEVICES={cuda_visible_devices}"
+            if len(args.include) or len(
+                    args.exclude) or args.num_nodes > 1 or args.num_gpus > 0:
+                print(
+                    f"{detected_str} but ignoring it because one or several of --include/--exclude/--num_gpus/--num_nodes cl args were used. If you want to use CUDA_VISIBLE_DEVICES don't pass any of these arguments to deepspeed."
+                )
+            else:
+                args.include = f"localhost:{cuda_visible_devices}"
+                print(f"{detected_str}: setting --include={args.include}")
+            del os.environ["CUDA_VISIBLE_DEVICES"]
+
+        if args.num_nodes >= 0 or args.num_gpus >= 0:
+            if args.include != "" or args.exclude != "":
+                raise ValueError("Cannot specify num_nodes/gpus with include/exclude")
 
     multi_node_exec = True
     if not resource_pool:
         resource_pool = {}
-        device_count = torch.cuda.device_count()
+        if torch.cuda.is_available():
+            device_count = torch.cuda.device_count()
+        else:
+            device_count = os.cpu_count()
```

Comment:
I changed this to choose the number of CPU cores based on either:

- Using the new parameter `args.num_cpu_cores` that users can explicitly set
- Detecting `torch.get_num_threads()` so that users can choose the number of cores with env vars `OMP_NUM_THREADS` or `MKL_NUM_THREADS`. Thoughts?